### PR TITLE
add get/api/reviews/id/comments, error handling and relavent tests

### PR DESCRIPTION
**Pull request by BBSharps**

### CODE REVIEW

File: __tests__/api-paths.test.js [hunk 1]
```diff
@@ -148,7 +148,66 @@ describe("GET /api/reviews/:review_id", () => {
...
 
-describe("error handeling", () => {
+describe("GET /api/reviews/:review_id/comments", () => {
+  test("status: 200 returns an array with the comment objects", () => {
+    return request(app)
+      .get("/api/reviews/3/comments")
+      .expect(200)
+      .then((data) => {
+        const reviews = data.body.comments;
+        console.log(reviews);
```
Comment (roisinoreilly) [inline __tests__/api-paths.test.js]: make sure you remove any console.logs before pushing

File: __tests__/api-paths.test.js [hunk 2]
```diff
@@ -148,7 +148,66 @@ describe("GET /api/reviews/:review_id", () => {
...
+        reviews.forEach((review) => {
```
Comment (roisinoreilly) [inline __tests__/api-paths.test.js]: if we're iterating through an array we first need to check the length to make sure it's not empty or we will get a false positive test

File: __tests__/api-paths.test.js [hunk 3]
```diff
@@ -148,7 +148,66 @@ describe("GET /api/reviews/:review_id", () => {
...
+    return request(app)
+      .get("/api/reviews/3/comments")
+      .expect(200)
+      .then((data) => {
+        const reviews = data.body.comments;
+        expect(reviews).toBeSortedBy("created_at", { descending: true });
+        expect(reviews).toHaveLength(3);
+      });
+  });
+  test("status: 200,if a valid id is given but there are no comments", () => {
```
Comment (roisinoreilly) [inline __tests__/api-paths.test.js:180]: nice edge case

File: db/api/controllers/index.js [hunk 4]
```diff
@@ -23,3 +27,19 @@ exports.getReviews = (req, res, next) => {
       next(err);
     });
 };
+
+exports.getReviewIdComments = (req, res, next) => {
+  const review_id = req.params.review_id;
+  Promise.all([selectReviews(review_id), selectReviewComments(review_id)])
```
Comment (roisinoreilly) [inline db/api/controllers/index.js]: why is this not rejecting

File: db/api/controllers/index.js [hunk 5]
```diff
@@ -23,3 +27,19 @@ exports.getReviews = (req, res, next) => {
...
+    .then((results) => {
+      if (results[0].length === 0)
+        return Promise.reject({ status: 404, msg: "not a valid id" });
+      const validComments = results[1].filter((comment) => {
```
Comment (roisinoreilly) [inline db/api/controllers/index.js]: you're having to look at results[1] because you're calling promise.all when you don't need to be, the tables don't need to be joined

File: db/api/controllers/index.js [hunk 6]
```diff
@@ -23,3 +27,19 @@ exports.getReviews = (req, res, next) => {
...
+        if (comment.comment_id !== null) return comment;
```
Comment (roisinoreilly) [inline db/api/controllers/index.js]: this shouldn't be happening it should be rejecting by itself

File: db/api/models/index.js [hunk 7]
```diff
@@ -38,3 +38,18 @@ exports.selectReviews = (review_id) => {
     return response.rows;
   });
 };
+
+exports.selectReviewComments = (review_id) => {
+  return db
+    .query(
+      `SELECT comment_id,comments.votes,comments.created_at,author,body, comments.review_id FROM reviews
+  LEFT JOIN comments ON reviews.review_id = comments.review_id
```
Comment (roisinoreilly) [inline db/api/models/index.js]: these tables dont need joining

File: __tests__/api-paths.test.js [hunk 8]
```diff
@@ -148,7 +148,66 @@ describe("GET /api/reviews/:review_id", () => {
...
 });
 
-describe("error handeling", () => {
+describe("GET /api/reviews/:review_id/comments", () => {
+  test("status: 200 returns an array with the comment objects", () => {
+    return request(app)
+      .get("/api/reviews/3/comments")
+      .expect(200)
+      .then((data) => {
+        const reviews = data.body.comments;
```
Comment (roisinoreilly) [inline __tests__/api-paths.test.js:155]: if we're getting back comments your variable should be named comments not reviews

File: __tests__/api-paths.test.js [hunk 9]
```diff
@@ -148,7 +148,66 @@ describe("GET /api/reviews/:review_id", () => {
...
+  test("status: 200,if a valid id is given but there are no comments", () => {
+    return request(app)
+      .get("/api/reviews/1/comments")
+      .expect(200)
+      .then((data) => {
+        const reviews = data.body.comments;
+        expect(reviews).toHaveLength(0);
+      });
+  });
+  test("status:404 when serching for a missing id in reviews", () => {
```
Comment (roisinoreilly) [inline __tests__/api-paths.test.js:189]: valid but non-existent id

File: __tests__/api-paths.test.js (hunk 8 above)
Comment (roisinoreilly) [inline __tests__/api-paths.test.js:155]: this variable should be comments because that's what we're getting back from the body

File: db/api/api.js [hunk 10]
```diff
@@ -4,12 +4,17 @@ const {
...
+const {
+  getCategories,
+  getReviews,
+  getReviewIdComments,
+} = require("./controllers/index");
 const app = express();
 
 app.get("/api/categories", getCategories);
 app.get("/api/reviews", getReviews);
 app.get("/api/reviews/:review_id", getReviews);
```
Comment (roisinoreilly) [inline db/api/api.js]: definitely should be two different controllers because we're getting back two different responses

File: db/api/controllers/index.js (hunk 6 above)
Comment (roisinoreilly) [reply db/api/controllers/index.js]: make sure we're not getting back null here and just filtering it out

**Review by roisinoreilly — CHANGES_REQUESTED** [review]
we've discussed it but to recap:
- getReviews from ticket 4 should respond with an array of reviews and doesn't need to take the review id as a parameter
- getReviewsByID from ticket 5 should take the id as a parameter but is a separate query to the database and needs its own model and controller, it responds with a single review object not an array
- your Promise.all should be handling your rejections you shouldn't need conditional logic to filter out valid comments - go back to your query and where you're joining tables
- be careful about variable names and be clear on what data you're receiving
- not necessary at all but you might find it easier to split your models and controllers into different files based on what part of the database they're accessing
- there's definitely an easier way to count the comments without using coalesce (however if it's working it's not that important)
- your models and controllers should ideally be in the root folder not in the db, but don't worry about it until everything else is fixed and be wary of your auto imports

drop us a HD and we'll go through it when you're ready, and go back through tickets 4 and 5 and make sure your responses are what the tickets are asking for. Your testing is good I can tell you just got mixed up, keep at it

File: api/controllers/index.js [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+      if (reviewId.length === 0) {
+        return Promise.reject({ status: 404, msg: "not a valid id" });
+      } else {
+        res.send({ reviewId: reviewId[0] });
+      }
+    })
+    .catch((err) => {
+      next(err);
+    });
+};
```
Comment (roisinoreilly) [inline api/controllers/index.js:43]: much better!!

File: __tests__/api-paths.test.js [hunk 12]
```diff
@@ -103,52 +103,108 @@ describe("GET /api/reviews/:review_id", () => {
...
+          })
+        );
+      });
+  });
+  test("should return only the required review", () => {
+    return request(app)
+      .get("/api/reviews/4")
+      .expect(200)
+      .then((data) => {
+        const result = data.body.reviewId;
```
Comment (roisinoreilly) [inline __tests__/api-paths.test.js:127]: not the most descriptive variable name

**Review by roisinoreilly — APPROVED** [review]
so much better, really great job implementing all the feedback it was a big job. let us know if you want to go through any of the queries again, otherwise merge when ready! 🎉
